### ISSUE TITLE
[Bug]: Cannot connect VMQ Diversity to Atlas 

### ISSUE THREAD

**Issue by TTia** [vernemq/vernemq#2323]
### Environment

- VerneMQ Version: vernemq/vernemq:2.0.1
- OS: Docker image
- Erlang/OTP version (if building from source): -
- Cluster size/standalone: both on cluster and on standalone instance

### Current Behavior

VerneMQ fails to connect to Atlas using the VMQ Diversity plugin for ACL:

- we are integrating VerneMQ (v2.0.1)
- with a MongoDB cluster (v5.0.25) hosted in Atlas.

When connecting to Atlas we configured the `vmq_diversity.mongodb.srv` property, instead of targeting the single cluster node through the `vmq_diversity.mongodb.host`. We are able to connect to that Atlas cluster in many ways, with different users (we tried mongosh, MongoDB Studio 3T, ...).

We tried to solve the issue both running the VerneMQ in GKE and locally.

### Expected behaviour

As it happens with the standalone MongoDB integration (using the `vmq_diversity.mongodb.host`), we expect to integrate an Atlas cluster with VerneMQ.

### Configuration, logs, error output, etc.

This is our current configuration (I've omitted the credentials and the mongodb SRV hostname):

```markdown
plugins.vmq_passwd = off
plugins.vmq_acl = off
plugins.vmq_diversity = on
vmq_diversity.auth_mongodb.enabled = on
vmq_diversity.mongodb.srv = db-name.asd123.mongodb.net
vmq_diversity.mongodb.database = connect_dev
vmq_diversity.mongodb.login = ***
vmq_diversity.mongodb.password = ***
```

Once started, VerneMQ tries to connect, it fails and start prompting the following errors over and over:

```
2024-06-26 10:48:05  __      ____  __  ____         _____ _______      ________ _____   _____ _____ _________     __ 
2024-06-26 10:48:05  \ \    / /  \/  |/ __ \       |  __ \_   _\ \    / /  ____|  __ \ / ____|_   _|__   __\ \   / / 
2024-06-26 10:48:05   \ \  / /| \  / | |  | |______| |  | || |  \ \  / /| |__  | |__) | (___   | |    | |   \ \_/ /  
2024-06-26 10:48:05    \ \/ / | |\/| | |  | |______| |  | || |   \ \/ / |  __| |  _  / \___ \  | |    | |    \   /   
2024-06-26 10:48:05     \  /  | |  | | |__| |      | |__| || |_   \  /  | |____| | \ \ ____) |_| |_   | |     | |    
2024-06-26 10:48:05      \/   |_|  |_|\___\_\      |_____/_____|   \/   |______|_|  \_\_____/|_____|  |_|     |_|    
2024-06-26 10:48:05                                                                                                  
2024-06-26 10:48:05 2024-06-26T08:48:05.049158+00:00 [info] <0.568.0> vmq_diversity_app:start/2:85: enable auth script for mongodb "./share/lua/auth/mongodb.lua"
2024-06-26 10:48:05 2024-06-26T08:48:05.127330+00:00 [error] <0.591.0> gen_server:error_info/8:1391: Generic server <0.591.0> terminating. Reason: tcp_closed. Last message: {tcp_closed,#Port<0.20>}. State: {state,#Port<0.20>,#{1 => #Fun<mc_worker_logic.0.40469528>},<<>>,{conn_state,safe,master,<<"connect_dev">>},#Ref<0.2004465639.1266155523.131779>,#Fun<mc_worker.0.103729946>,gen_tcp}.
2024-06-26 10:48:05 2024-06-26T08:48:05.127541+00:00 [error] <0.591.0> proc_lib:crash_report/4:584: crasher: initial call: mc_worker:init/1, pid: <0.591.0>, registered_name: [], exit: {tcp_closed,[{gen_server,handle_common_reply,8,[{file,"gen_server.erl"},{line,1226}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}, ancestors: [<0.583.0>,<0.580.0>,auth_mongodb,vmq_diversity_sup,<0.563.0>], message_queue_len: 0, messages: [], links: [<0.583.0>], dictionary: [{rand_seed,{#{type => exsss,next => #Fun<rand.0.65977474>,bits => 58,uniform => #Fun<rand.1.65977474>,uniform_n => #Fun<rand.2.65977474>,jump => #Fun<rand.3.65977474>},[110029574451504855|285073100845245618]}}], trap_exit: false, status: running, heap_size: 4185, stack_size: 28, reductions: 11498; neighbours:
2024-06-26 10:48:05 2024-06-26T08:48:05.127799+00:00 [error] <0.583.0> gen_server:error_info/8:1391: Generic server <0.583.0> terminating. Reason: {{tcp_closed,{gen_server,call,[<0.591.0>,{query,undefined,<<"$cmd">>,false,false,false,false,false,0,-1,{<<"buildinfo">>,1},#{}},infinity]}},[{gen_server,call,3,[{file,"gen_server.erl"},{line,419}]},{mc_connection_man,request_worker,2,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/connection/mc_connection_man.erl"},{line,95}]},{mc_connection_man,read_one,2,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/connection/mc_connection_man.erl"},{line,37}]},{mc_connection_man,command,2,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/connection/mc_connection_man.erl"},{line,46}]},{mc_worker_api,get_version,1,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,85}]},{mc_worker_api,connect,1,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,67}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/opt/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]}]}. Last message: connect. State: {state,mongodb,#Fun<vmq_diversity_sup.4.29008357>,#Fun<vmq_diversity_sup.5.29008357>,undefined,false,1000,undefined}.
2024-06-26 10:48:05 2024-06-26T08:48:05.128097+00:00 [error] <0.583.0> proc_lib:crash_report/4:584: crasher: initial call: vmq_diversity_worker_wrapper:init/1, pid: <0.583.0>, registered_name: [], exit: {{tcp_closed,{gen_server,call,[<0.591.0>,{query,undefined,<<"$cmd">>,false,false,false,false,false,0,-1,{<<"buildinfo">>,1},#{}},infinity]}},[{gen_server,call,3,[{file,"gen_server.erl"},{line,419}]},{mc_connection_man,request_worker,2,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/connection/mc_connection_man.erl"},{line,95}]},{mc_connection_man,read_one,2,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/connection/mc_connection_man.erl"},{line,37}]},{mc_connection_man,command,2,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/connection/mc_connection_man.erl"},{line,46}]},{mc_worker_api,get_version,1,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,85}]},{mc_worker_api,connect,1,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,67}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/opt/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]}]}, ancestors: [<0.580.0>,auth_mongodb,vmq_diversity_sup,<0.563.0>], message_queue_len: 1, messages: [{'EXIT',<0.591.0>,tcp_closed}], links: [<0.580.0>,<0.579.0>], dictionary: [], trap_exit: true, status: running, heap_size: 17731, stack_size: 28, reductions: 24740; neighbours:
```
```


### Code of Conduct

- [X] I agree to follow the VerneMQ's Code of Conduct

**Comment by ioolkos** [comment]
Your config opens a TCP connection, is the remote endpoint also TCP, or TLS?

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
Thanks for the reply @ioolkos. Yes, we are using TLS but I don't see any configuration regarding TLS/SSL for the integration with Mongo (on contrary, I can see the SSL toggle for PostgreSQL or CockroachDB).

With `mongosh` I'm not providing any certificates and I'm able to connect.

https://docs.vernemq.com/configuring-vernemq/db-auth#cockroachdb

**Comment by ioolkos** [comment]
Full mappings are here: https://github.com/vernemq/vernemq/blob/main/apps/vmq_diversity/priv/vmq_diversity.schema

try
```
vmq_diversity.mongodb.ssl = on
```
or

```
vmq_diversity.mongodb.ssl = on
vmq_diversity.mongodb.cafile = path/to/cacerts.pem
```

That said, I just realised we might have to extend SSL support in the driver here due to changes in OTP 26. (as, for example, done for the Postgres driver).
But let's see.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
I've enabled `vmq_diversity.mongodb.ssl = on` but I did not provide a certificate, do you think it's needed? While connecting  with `mongosh` it's not required.

Anyway, enabling the `ssl` triggers the following errors:

```
2024-06-26 15:46:48  __      ____  __  ____         _____ _______      ________ _____   _____ _____ _________     __ 
2024-06-26 15:46:48  \ \    / /  \/  |/ __ \       |  __ \_   _\ \    / /  ____|  __ \ / ____|_   _|__   __\ \   / / 
2024-06-26 15:46:48   \ \  / /| \  / | |  | |______| |  | || |  \ \  / /| |__  | |__) | (___   | |    | |   \ \_/ /  
2024-06-26 15:46:48    \ \/ / | |\/| | |  | |______| |  | || |   \ \/ / |  __| |  _  / \___ \  | |    | |    \   /   
2024-06-26 15:46:48     \  /  | |  | | |__| |      | |__| || |_   \  /  | |____| | \ \ ____) |_| |_   | |     | |    
2024-06-26 15:46:48      \/   |_|  |_|\___\_\      |_____/_____|   \/   |______|_|  \_\_____/|_____|  |_|     |_|    
2024-06-26 15:46:48                                                                                                  
2024-06-26 15:46:48 2024-06-26T13:46:48.093365+00:00 [info] <0.568.0> vmq_diversity_app:start/2:85: enable auth script for mongodb "./share/lua/auth/mongodb.lua"
2024-06-26 15:46:48 2024-06-26T13:46:48.133172+00:00 [error] <0.583.0> gen_server:error_info/8:1391: Generic server <0.583.0> terminating. Reason: {{badmatch,{error,{options,incompatible,[{verify,verify_peer},{cacerts,undefined}]}}},[{mc_worker_api,connect,1,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,61}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/opt/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}. Last message: connect. State: {state,mongodb,#Fun<vmq_diversity_sup.4.29008357>,#Fun<vmq_diversity_sup.5.29008357>,undefined,false,1000,undefined}.
2024-06-26 15:46:48 2024-06-26T13:46:48.133169+00:00 [error] <0.584.0> gen_server:error_info/8:1391: Generic server <0.584.0> terminating. Reason: {{badmatch,{error,{options,incompatible,[{verify,verify_peer},{cacerts,undefined}]}}},[{mc_worker_api,connect,1,[{file,"/opt/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,61}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/opt/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}. Last message: connect. State: {state,mongodb,#Fun<vmq_diversity_sup.4.29008357>,#Fun<vmq_diversity_sup.5.29008357>,undefined,false,1000,undefined}.
```

**Comment by ioolkos** [comment]
Ok, this already proves the suspicion. A release built on OTP 25 would work here. (feel free to try)

OTP 26 changes the `verify` option to `verify_peer`. To do that the client (which is VerneMQ here) needs a CAfile. That's why the Erlang TLS stack tells us here that the options are incompatible.

The default for `verify` on OTP 25 is `verify_none`, so this is not enforced.

Mind you, even with a working CA chain, I bet the next thing you will see is issues with AWS wildcard certs. For which we need to pass a hostname_verification function.

In other words: We need to do similar adaptations as for the Postgres plugin.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
Sorry for the "dumb" question, but I'm not familiar with the VerneMQ internals and I 've never built the project from scratch. How can I upgrade Erlang to OTP 25? Is this possible from the docker-image perspective?

**Comment by ioolkos** [comment]
@TTia can't type up a full introduction here.

You need Erlang on your system (Linux).
Then
https://github.com/vernemq/vernemq?tab=readme-ov-file#building-vernemq
https://github.com/vernemq/vernemq?tab=readme-ov-file#starting-vernemq

Also note the potential fix (for OTP 26) in the PR I just opened.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
We have built the from #2324 using OTP 25 (https://packages.debian.org/sid/erlang-dev) and built a new docker image using the latest changes.

Then we applied the following conf, as suggested:

```
      [...]
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__VERIFY=verify_peer
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__USE_SYSTEM_CAS=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__CUSTOMIZE_HOSTNAME_CHECK=https
```

It is failing with the following error:

```
2024-06-27 11:50:34  __      ____  __  ____         _____ _______      ________ _____   _____ _____ _________     __ 
2024-06-27 11:50:34  \ \    / /  \/  |/ __ \       |  __ \_   _\ \    / /  ____|  __ \ / ____|_   _|__   __\ \   / / 
2024-06-27 11:50:34   \ \  / /| \  / | |  | |______| |  | || |  \ \  / /| |__  | |__) | (___   | |    | |   \ \_/ /  
2024-06-27 11:50:34    \ \/ / | |\/| | |  | |______| |  | || |   \ \/ / |  __| |  _  / \___ \  | |    | |    \   /   
2024-06-27 11:50:34     \  /  | |  | | |__| |      | |__| || |_   \  /  | |____| | \ \ ____) |_| |_   | |     | |    
2024-06-27 11:50:34      \/   |_|  |_|\___\_\      |_____/_____|   \/   |______|_|  \_\_____/|_____|  |_|     |_|    
2024-06-27 11:50:34                                                                                                  
2024-06-27 11:50:34 2024-06-27T09:50:34.222753+00:00 [info] <0.614.0> vmq_diversity_app:start/2:85: enable auth script for mongodb "./share/lua/auth/mongodb.lua"
2024-06-27 11:50:34 2024-06-27T09:50:34.352002+00:00 [error] <0.632.0> gen_server:error_info/8:1391: Generic server <0.632.0> terminating. Reason: {{badmatch,{error,{options,{server_name_indication,localhost}}}},[{mc_worker_api,connect,1,[{file,"/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,61}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}. Last message: connect. State: {state,mongodb,#Fun<vmq_diversity_sup.4.29008357>,#Fun<vmq_diversity_sup.5.29008357>,undefined,false,1000,undefined}.
```

**Comment by ioolkos** [comment]
The server name will be taken from
```
vmq_diversity.mongodb.ssl.sni = your-server-name-indication
```
and if that is not set, it will be taken from `vmq_diversity.mongodb.host` (which seems to be localhost here, as the default)

Can you try setting the sni over the config, to something that makes sense (maybe it could be the same name as the `srv`, not sure about that)


---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
We have built the from #2324 using OTP 26 and built a new docker image using the latest changes.

Then we applied the following conf, as suggested:

```
      [...]
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__VERIFY=verify_peer
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__USE_SYSTEM_CAS=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__CUSTOMIZE_HOSTNAME_CHECK=https
```

It is failing with the following error:

```
2024-06-27 11:50:34  __      ____  __  ____         _____ _______      ________ _____   _____ _____ _________     __ 
2024-06-27 11:50:34  \ \    / /  \/  |/ __ \       |  __ \_   _\ \    / /  ____|  __ \ / ____|_   _|__   __\ \   / / 
2024-06-27 11:50:34   \ \  / /| \  / | |  | |______| |  | || |  \ \  / /| |__  | |__) | (___   | |    | |   \ \_/ /  
2024-06-27 11:50:34    \ \/ / | |\/| | |  | |______| |  | || |   \ \/ / |  __| |  _  / \___ \  | |    | |    \   /   
2024-06-27 11:50:34     \  /  | |  | | |__| |      | |__| || |_   \  /  | |____| | \ \ ____) |_| |_   | |     | |    
2024-06-27 11:50:34      \/   |_|  |_|\___\_\      |_____/_____|   \/   |______|_|  \_\_____/|_____|  |_|     |_|    
2024-06-27 11:50:34                                                                                                  
2024-06-27 11:50:34 2024-06-27T09:50:34.222753+00:00 [info] <0.614.0> vmq_diversity_app:start/2:85: enable auth script for mongodb "./share/lua/auth/mongodb.lua"
2024-06-27 11:50:34 2024-06-27T09:50:34.352002+00:00 [error] <0.632.0> gen_server:error_info/8:1391: Generic server <0.632.0> terminating. Reason: {{badmatch,{error,{options,{server_name_indication,localhost}}}},[{mc_worker_api,connect,1,[{file,"/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,61}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}. Last message: connect. State: {state,mongodb,#Fun<vmq_diversity_sup.4.29008357>,#Fun<vmq_diversity_sup.5.29008357>,undefined,false,1000,undefined}.
```

**Comment by ioolkos** [comment]
Also note that you can emulate OTP 25 default behaviour (not verifying the peer) by:

```
vmq_diversity.mongodb.ssl.verify = verify_none
```
---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
Setting the `DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__SNI` as the `DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SRV` is causing VerneMQ raising the following error:

```
vernemq-1  | 2024-06-27T12:54:04.580473+00:00 [error] <0.1099.0> proc_lib:crash_report/4:584: crasher: initial call: vmq_diversity_worker_wrapper:init/1, pid: <0.1099.0>, registered_name: [], error: {{badmatch,{error,{options,{server_name_indication,'db-name.asd123.mongodb.net'}}}},[{mc_worker_api,connect,1,[{file,"/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,61}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}, ancestors: [<0.626.0>,auth_mongodb,vmq_diversity_sup,<0.609.0>], message_queue_len: 1, messages: [{'EXIT',<0.1100.0>,normal}], links: [<0.626.0>,<0.625.0>], dictionary: [], trap_exit: true, status: running, heap_size: 6772, stack_size: 28, reductions: 15717; ^C,[{mc_worker_api,connect,1,[{file,"/vernemq/_build/default/lib/mongodb/src/api/mc_worker_api.erl"},{line,61}]},{vmq_diversity_worker_wrapper,handle_info,2,[{file,"/vernemq/apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl"},{line,176}]},{gen_server,try_handle_info,3,[{file,"gen_server.erl"},{line,1095}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}. Last message: connect. State: {state,mongodb,#Fun<vmq_diversity_sup.4.29008357>,#Fun<vmq_diversity_sup.5.29008357>,undefined,false,1000,undefined}.
```

By the way, with `mongosh` I'm able to connect using TLS without specify any further parameter:

```
mongosh 'mongodb+srv://db-name.asd123.mongodb.net' --username *** --password *** --authenticationDatabase admin --tls
```

Here the full config:

```
services:
  vernemq:
    # image: vernemq/vernemq:2.0.1
    image: <custom build image>
    environment:
      - DOCKER_VERNEMQ_ACCEPT_EULA=yes
      - DOCKER_VERNEMQ_PLUGINS__VMQ_PASSWD=off
      - DOCKER_VERNEMQ_PLUGINS__VMQ_ACL=off
      - DOCKER_VERNEMQ_PLUGINS__VMQ_DIVERSITY=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__AUTH_MONGODB__ENABLED=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SRV=db-name.asd123.mongodb.net
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__DATABASE=***
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__LOGIN=***
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__PASSWORD=***

      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__SNI=db-name.asd123.mongodb.net
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__VERIFY=verify_peer
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__USE_SYSTEM_CAS=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__MONGODB__SSL__CUSTOMIZE_HOSTNAME_CHECK=https
    ports:
      - 2883:1883
```

**Comment by ioolkos** [comment]
Maybe you can inquire with mongodb.net what the correct SNI is. Just a domain name?
I can also take out the automated forwarding of SNI (as part of the SSL options) to the MongoDB driver library, although I'm not convinced this would be the correct approach.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by ioolkos** [comment]
@TTia Did you have a chance to further test this?

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by ioolkos** [comment]
I'd like to include this update in the upcoming `2.0.2` bugfix release. Thanks if anyone can test this with Atlas specifically.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by TTia** [comment]
Hi @oluigipo , we didn't resolve the issue nor investigate this further: we switched to postgres.

**Comment by ioolkos** [comment]
@oluigipo, @TTia Thanks for testing against Atlas! Should you have additional insights that would let us jointly work towards a fix, let us know. I'm re-opening the issue.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
